### ISSUE TITLE
crash

### ISSUE THREAD

**Issue by matthiaskrgr** [clangen/musikcube#151]
Not sure what exactly triggers this, it seems to happen randomly while skipping from file to file

````
frame 0, skipping ID3 tag (4083 bytes)

  frames: 7733, br: 128017 b/s, sr: 44100 Hz, ch: 2, layer: 3, joint stereo: 0
  dual channel: 0, vbr: 0, duration: 202.004 s, xing: 1
v2.3.0
frame 7733, skipping ID3 tag (4083 bytes)
AlsaOut: closing PCM handle

  frames: 10589, br: 192000 b/s, sr: 44100 Hz, ch: 2, layer: 3, joint stereo: 0
  dual channel: 0, vbr: 0, duration: 276.61 s, xing: 0
AlsaOut: error! File descriptor in bad state
AlsaOut: device seems to be prepared for use!
AlsaOut: device format initialized from buffer

  frames: 12920, br: 192000 b/s, sr: 44100 Hz, ch: 2, layer: 3, joint stereo: 0
  dual channel: 0, vbr: 0, duration: 337.502 s, xing: 0
v2.3.0
frame 0, skipping ID3 tag (59392 bytes)
AlsaOut: closing PCM handle
AlsaOut: error! File descriptor in bad state

  frames: 8795, br: 190155 b/s, sr: 44100 Hz, ch: 2, layer: 3, joint stereo: 1
  dual channel: 0, vbr: 1, duration: 229.746 s, xing: 1
v2.3.0
frame 8795, skipping ID3 tag (59392 bytes)
AlsaOut: device seems to be prepared for use!
AlsaOut: device format initialized from buffer
v2.3.0
frame 0, skipping ID3 tag (59392 bytes)

  frames: 12614, br: 187718 b/s, sr: 44100 Hz, ch: 2, layer: 3, joint stereo: 1
  dual channel: 0, vbr: 1, duration: 329.508 s, xing: 1
v2.3.0
frame 12614, skipping ID3 tag (59392 bytes)
AlsaOut: closing PCM handle
AlsaOut: error! File descriptor in bad state
AlsaOut: device seems to be prepared for use!
AlsaOut: device format initialized from buffer
AlsaOut: closing PCM handle
AlsaOut: error! File descriptor in bad state
/home/matthias/vcs/github/musikcube/src/plugins/m4adecoder/M4aDecoder.cpp:225:21: runtime error: null pointer passed as argument 1, which is declared to never be null
/usr/include/string.h:43:28: note: nonnull attribute specified here
    #0 0x7f29a84cfd0d in M4aDecoder::GetBuffer(musik::core::sdk::IBuffer*) /home/matthias/vcs/github/musikcube/src/plugins/m4adecoder/M4aDecoder.cpp:224:17
    #1 0x18a14a4 in musik::core::audio::Stream::GetNextBufferFromDecoder() /home/matthias/vcs/github/musikcube/src/core/audio/Stream.cpp:155:25
    #2 0x189e0cf in musik::core::audio::Stream::RefillInternalBuffers() /home/matthias/vcs/github/musikcube/src/core/audio/Stream.cpp:230:18
    #3 0x189d0a2 in musik::core::audio::Stream::OpenStream(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /home/matthias/vcs/github/musikcube/src/core/audio/Stream.cpp:135:19
    #4 0x18fb9d6 in musik::core::audio::playerThreadLoop(musik::core::audio::Player*) /home/matthias/vcs/github/musikcube/src/core/audio/Player.cpp:288:25
    #5 0x191edcc in void std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:933:11
    #6 0x191e884 in void std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)>::operator()<, void>() /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:991:17
    #7 0x191e5e5 in void std::_Bind_simple<std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)> ()>::_M_invoke<>(std::_Index_tuple<>) /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:1390:18
    #8 0x191e470 in std::_Bind_simple<std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)> ()>::operator()() /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:1380:16
    #9 0x191d78d in std::thread::_State_impl<std::_Bind_simple<std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)> ()> >::_M_run() /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/thread:197:13
    #10 0x7f29afa9076e  (/lib64/libstdc++.so.6+0xbb76e)
    #11 0x7f29b03cd739 in start_thread (/lib64/libpthread.so.0+0x7739)
    #12 0x7f29aefeee0e in __GI___clone (/lib64/libc.so.6+0x107e0e)

SUMMARY: AddressSanitizer: undefined-behavior /home/matthias/vcs/github/musikcube/src/plugins/m4adecoder/M4aDecoder.cpp:225:21 in 
/home/matthias/vcs/github/musikcube/src/core/audio/Buffer.cpp:149:9: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:43:28: note: nonnull attribute specified here
    #0 0x113363a in musik::core::audio::Buffer::Copy(float*, long, long) /home/matthias/vcs/github/musikcube/src/core/audio/Buffer.cpp:149:9
    #1 0x189ffcd in musik::core::audio::Stream::RefillInternalBuffers() /home/matthias/vcs/github/musikcube/src/core/audio/Stream.cpp:274:21
    #2 0x189d0a2 in musik::core::audio::Stream::OpenStream(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /home/matthias/vcs/github/musikcube/src/core/audio/Stream.cpp:135:19
    #3 0x18fb9d6 in musik::core::audio::playerThreadLoop(musik::core::audio::Player*) /home/matthias/vcs/github/musikcube/src/core/audio/Player.cpp:288:25
    #4 0x191edcc in void std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:933:11
    #5 0x191e884 in void std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)>::operator()<, void>() /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:991:17
    #6 0x191e5e5 in void std::_Bind_simple<std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)> ()>::_M_invoke<>(std::_Index_tuple<>) /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:1390:18
    #7 0x191e470 in std::_Bind_simple<std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)> ()>::operator()() /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/functional:1380:16
    #8 0x191d78d in std::thread::_State_impl<std::_Bind_simple<std::_Bind<void (*(musik::core::audio::Player*))(musik::core::audio::Player*)> ()> >::_M_run() /usr/lib/gcc/x86_64-redhat-linux/6.4.1/../../../../include/c++/6.4.1/thread:197:13
    #9 0x7f29afa9076e  (/lib64/libstdc++.so.6+0xbb76e)
    #10 0x7f29b03cd739 in start_thread (/lib64/libpthread.so.0+0x7739)
    #11 0x7f29aefeee0e in __GI___clone (/lib64/libc.so.6+0x107e0e)

SUMMARY: AddressSanitizer: undefined-behavior /home/matthias/vcs/github/musikcube/src/core/audio/Buffer.cpp:149:9 in 
AlsaOut: device seems to be prepared for use!
AlsaOut: device format initialized from buffer
AlsaOut: closing PCM handle
ASAN:DEADLYSIGNAL
=================================================================
==4308==ERROR: AddressSanitizer: SEGV on unknown address 0x611033000064 (pc 0x7f299fed58da bp 0x000000000800 sp 0x7f2998aab518 T4)
==4308==The signal is caused by a READ memory access.
    #0 0x7f299fed58d9  (/lib64/libasound.so.2+0x948d9)
    #1 0x7f299fe96b20  (/lib64/libasound.so.2+0x55b20)
    #2 0x7f299fed6927  (/lib64/libasound.so.2+0x95927)
    #3 0x7f29a8295e88 in AlsaOut::WriteLoop() /home/matthias/vcs/github/musikcube/src/plugins/alsaout/AlsaOut.cpp:420:17
    #4 0x7f29a82c0edd in boost::_mfi::mf0<void, AlsaOut>::operator()(AlsaOut*) const /usr/include/boost/bind/mem_fn_template.hpp:49:29
    #5 0x7f29a82c0c16 in void boost::_bi::list1<boost::_bi::value<AlsaOut*> >::operator()<boost::_mfi::mf0<void, AlsaOut>, boost::_bi::list0>(boost::_bi::type<void>, boost::_mfi::mf0<void, AlsaOut>&, boost::_bi::list0&, int) /usr/include/boost/bind/bind.hpp:259:9
    #6 0x7f29a82c0927 in boost::_bi::bind_t<void, boost::_mfi::mf0<void, AlsaOut>, boost::_bi::list1<boost::_bi::value<AlsaOut*> > >::operator()() /usr/include/boost/bind/bind.hpp:1222:16
    #7 0x7f29a82be76c in boost::detail::thread_data<boost::_bi::bind_t<void, boost::_mfi::mf0<void, AlsaOut>, boost::_bi::list1<boost::_bi::value<AlsaOut*> > > >::run() /usr/include/boost/thread/detail/thread.hpp:116:17
    #8 0x7f29b1092145  (/lib64/libboost_thread.so.1.60.0+0x12145)
    #9 0x7f29b03cd739 in start_thread (/lib64/libpthread.so.0+0x7739)
    #10 0x7f29aefeee0e in __GI___clone (/lib64/libc.so.6+0x107e0e)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/lib64/libasound.so.2+0x948d9) 
Thread T4 created by T0 here:
    #0 0x8715d6 in pthread_create /home/matthias/LLVM/LLVM_4_0/stage_2/llvm/projects/compiler-rt/lib/asan/asan_interceptors.cc:305:3
    #1 0x7f29b1091498 in boost::thread::start_thread_noexcept() (/lib64/libboost_thread.so.1.60.0+0x11498)
    #2 0x7f29a82a6a4d in boost::thread::thread<boost::_bi::bind_t<void, boost::_mfi::mf0<void, AlsaOut>, boost::_bi::list1<boost::_bi::value<AlsaOut*> > > >(boost::_bi::bind_t<void, boost::_mfi::mf0<void, AlsaOut>, boost::_bi::list1<boost::_bi::value<AlsaOut*> > >&&) /usr/include/boost/thread/detail/thread.hpp:265:13
    #3 0x7f29a82933ff in AlsaOut::AlsaOut() /home/matthias/vcs/github/musikcube/src/plugins/alsaout/AlsaOut.cpp:146:33
    #4 0x7f29a829107d in GetAudioOutput /home/matthias/vcs/github/musikcube/src/plugins/alsaout/alsaout_plugin.cpp:61:13
    #5 0xf16988 in void musik::core::PluginFactory::QueryInterface<musik::core::sdk::IOutput, musik::core::PluginFactory::DestroyDeleter<musik::core::sdk::IOutput> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<void (std::shared_ptr<musik::core::sdk::IOutput>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) /home/matthias/vcs/github/musikcube/src/core/plugin/PluginFactory.h:90:41
    #6 0xf14adb in std::vector<std::shared_ptr<musik::core::sdk::IOutput>, std::allocator<std::shared_ptr<musik::core::sdk::IOutput> > > musik::core::PluginFactory::QueryInterface<musik::core::sdk::IOutput, musik::core::PluginFactory::DestroyDeleter<musik::core::sdk::IOutput> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/matthias/vcs/github/musikcube/src/core/plugin/PluginFactory.h:103:17
    #7 0xf06786 in musik::core::audio::outputs::GetAllOutputs() /home/matthias/vcs/github/musikcube/src/core/audio/Outputs.cpp:83:26
    #8 0xf074a2 in musik::core::audio::outputs::SelectedOutput() /home/matthias/vcs/github/musikcube/src/core/audio/Outputs.cpp:110:48
    #9 0x18e7bdc in musik::core::audio::GaplessTransport::GaplessTransport() /home/matthias/vcs/github/musikcube/src/core/audio/GaplessTransport.cpp:69:20
    #10 0x18b8452 in musik::glue::audio::MasterTransport::SwitchTo(musik::glue::audio::MasterTransport::Type) /home/matthias/vcs/github/musikcube/src/glue/audio/MasterTransport.cpp:78:43
    #11 0x18b6a3e in musik::glue::audio::MasterTransport::MasterTransport() /home/matthias/vcs/github/musikcube/src/glue/audio/MasterTransport.cpp:54:11
    #12 0x8b9b6d in main /home/matthias/vcs/github/musikcube/src/musikcube/Main.cpp:134:21
    #13 0x7f29aef07400 in __libc_start_main (/lib64/libc.so.6+0x20400)

==4308==ABORTING

````

**Comment by clangen** [comment]
Looks like this can happen if the M4A decoder returns a null buffer. I can't repro myself, so I'm going to add a null check to prevent the crash, but I'm not sure what the behavior will be. I think worst case scenario playback stalls, best case a subsequent read kicks things back into gear. 

Fix pushed to master, will be in the next release.

**Comment by clangen** [comment]
Crash should be fixed in 0.25.1. As said, I wasn't able to repro so I'm not sure what the actual behavior will be. If things are totally screwed up please re-open.
